src/core.c/HyperSeq.pm6

```diff
@@ -1,46 +1,59 @@
 # A HyperSeq performs batches of work in parallel, but retains order of output
 # values relative to input values.
 #?if !js
-my class HyperSeq does Iterable does Sequence {
+
+my role Sequence::Hyper does Iterable does Sequence {
     has HyperConfiguration $.configuration;
     has Rakudo::Internals::HyperWorkStage $!work-stage-head;
+    has $!joiner;
 
     submethod BUILD(:$!configuration!, :$!work-stage-head!) {}
 
-    method iterator(HyperSeq:D: --> Iterator) {
-        my $joiner := Rakudo::Internals::HyperToIterator.new:
+    method iterator(::?CLASS:D: --> Iterator) {
+        X::Seq::Consumed.new(:kind(::?CLASS.^name)).throw
+            if nqp::isconcrete($!joiner);
+        $!joiner := Rakudo::Internals::HyperToIterator.new:
```

Comment:
I wonder if using `cas` or even just a native int attribute and an atomic increment would be more robust here. That way we'd catch attempts to consume it in parallel too (and don't have to be concerned about whether we're extending the lifetime of something by storing it).

Reply:
With all my love for parallelizing it's the kind of error I tend to repeat. Too much Perl in the veins, perhaps. ;) Also good point on preserving the joiner when we don't really need it. Because it's a case of boolean semantics, cas over atomicint would do the best.

Thanks!